### DISCUSSION TITLE
Setting to change Default KDF

### DISCUSSION THREAD

**Discussion by iaburton** [dani-garcia/vaultwarden#3520] [Q&A]
Hello!

I searched through the [wiki](https://github.com/dani-garcia/vaultwarden/wiki) as well as the [env template](https://github.com/dani-garcia/vaultwarden/blob/main/.env.template) and didn't see if there was a way to set the default kdf alg for new users?

From what I understand currently, if a new user is created and imports or adds their passwords before changing from pbkdf2 to argon they would then subsequently need to change their (newly made) master password in order [for their encryption key to change](https://bitwarden.com/help/kdf-algorithms/).

Not that changing the master pass is totally necessary, I just thought it would be a good idea to allow a change in default so non-technical users don't need to concern themselves with it.

Now that I write this out though, I haven't added a new user in a bit. With the argon changes are you able to select your desired kdf during registration? Perhaps that would be sufficient.

If that already exists then please pardon me and this discussion :slightly_smiling_face:

**Comment by BlackDex** [comment]
If I'm correct, the default can't be changed. But i haven't looked into it that well.
But what i understood is that some clients still might have issues with there Argon2 implementation. And before they will make it a default, they want it more tested i think.

**Reply by iaburton** [reply]
Yeah I understand the issue with some perhaps outdated / older clients having issue with argon, and why it might not be the default on a new install or (for a) new user. Though I was hoping for a manual setting in the mean time, so current users can change their default on existing installs or let newly registered users pick before they add/import their data. Maybe a warning that argon isn't supported everywhere yet :shrug: just a thought.

Are you saying/thinking that vaultwarden is unlikely to do something like this in order to stay more in line with bitwarden? For example, wait it out until the latter makes it the default?

**Reply by BlackDex** [reply]
That is exactly what I'm saying.
We will not change fundamental functionality. Also, you can create accounts in other clients then just the web-vault. So, if, then it needs to be done for all clients, and that's not something we can control.

And there may be a reason, not known to us for not setting this as a default.

**Reply by iaburton** [reply]
>Also, you can create accounts in other clients then just the web-vault. So, if, then it needs to be done for all clients, and that's not something we can control.

Ah understood, forgot about that :sweat_smile: 

>We will not change fundamental functionality. 

So that includes a _setting_ to change the default? I'll update my discussion title as it may be misleading; the ask is for a setting to change the default, not to change the default itself to Argon. The default can be left PBKDF2.

**Reply by BlackDex** [reply]
That will not matter, it's a client side setting if I'm correct (though not looked into it detailed). So creating a server side setting will not help.

**Reply by iaburton** [reply]
Oh okay, I could be misunderstanding how it works but I thought the same alg was used on both sides. If this one is just client side though, I can see the hesitation on a change like this for the time being.

**Reply by stefan0xC** [reply]
As far as I've looked into it the default KDF settings for creating new accounts are hardcoded in the clients here: 
https://github.com/bitwarden/clients/blob/web-v2023.4.2/libs/common/src/enums/kdf-type.enum.ts

**Reply by BlackDex** [reply]
The server doesn't encrypt, decrypt or hashes anything in regards too ciphers.

It only hashes a binary hashed representation by default with a 600_000 rounds of PBKDF to the database. And further it uses Argon2 to be able to have a hashed admin password. All other items in regards to hashing and encryption is done client side.
